chronos/asyncloop.nim

```diff
@@ -425,59 +459,91 @@ when defined(windows):
 
   proc poll*() {.raises: [Defect, CatchableError].} =
     ## Perform single asynchronous step, processing timers and completing
-    ## unblocked tasks. Blocks until at least one event has completed.
+    ## tasks. Blocks until at least one event has completed.
     ##
     ## Exceptions raised here indicate that waiting for tasks to be unblocked
     ## failed - exceptions from within tasks are instead propagated through
     ## their respective futures and not allowed to interrrupt the poll call.
     let loop = getThreadDispatcher()
-    var curTime = Moment.now()
-    var curTimeout = DWORD(0)
-    var noNetworkEvents = false
+    var
+      curTime = Moment.now()
+      curTimeout = DWORD(0)
+      events: array[MaxEventsCount, OVERLAPPED_ENTRY]
+
+    # On reentrant `poll` calls from `processCallbacks`, e.g., `waitFor`,
+    # complete pending work of the outer `processCallbacks` call.
+    # On non-reentrant `poll` calls, this only removes sentinel element.
+    processCallbacks(loop)
 
     # Moving expired timers to `loop.callbacks` and calculate timeout
     loop.processTimersGetTimeout(curTimeout)
 
-    # Processing handles
-    var lpNumberOfBytesTransferred: DWORD
-    var lpCompletionKey: ULONG_PTR
-    var customOverlapped: PtrCustomOverlapped
-
-    let res = getQueuedCompletionStatus(
-      loop.ioPort, addr lpNumberOfBytesTransferred,
-      addr lpCompletionKey, cast[ptr POVERLAPPED](addr customOverlapped),
-      curTimeout).bool
-
-    if res:
-      customOverlapped.data.bytesCount = lpNumberOfBytesTransferred
-      customOverlapped.data.errCode = OSErrorCode(-1)
+    let networkEventsCount =
+      if isNil(loop.getQueuedCompletionStatusEx):
+        let res = getQueuedCompletionStatus(
+          loop.ioPort,
+          addr events[0].dwNumberOfBytesTransferred,
+          addr events[0].lpCompletionKey,
+          cast[ptr POVERLAPPED](addr events[0].lpOverlapped),
+          curTimeout
+        )
+        if res == WINBOOL(0):
+          let errCode = osLastError()
+          if not(isNil(events[0].lpOverlapped)):
+            1
+          else:
+            if int32(errCode) != WAIT_TIMEOUT:
+              raiseOSError(errCode)
+            0
+        else:
+          1
+      else:
+        var eventsReceived = ULONG(0)
+        let res = loop.getQueuedCompletionStatusEx(
+          loop.ioPort,
+          addr events[0],
+          ULONG(len(events)),
+          eventsReceived,
+          curTimeout,
+          WINBOOL(0)
+        )
+        if res == WINBOOL(0):
+          let errCode = osLastError()
+          if int32(errCode) != WAIT_TIMEOUT:
+            raiseOSError(errCode)
+          0
+        else:
+          eventsReceived
+
+    for i in 0 ..< networkEventsCount:
+      var customOverlapped = events[i].lpOverlapped
+      customOverlapped.data.errCode =
+        block:
+          let res = cast[uint64](customOverlapped.internal)
```

Comment:
Does regular `getQueuedCompletionStatus` also update `internal`/`status`? 
Previously, the return value was used to decide on putting -1 or `osLastError()` instead.

Reply:
Of course. OVERLAPPED mechanism is not part of IOCP actually. So it doesn't matter what function we are using to query IOCP port. And yes previous we have used `GetLastError()`, but after my checks i found that both sources returns identical information when OVERLAPPED structure pointer is not `NULL`.